src/ApplyFiltersDynamicFunctionReturnTypeExtension.php

```diff
@@ -0,0 +1,110 @@
+<?php
+
+/**
+ * Set return type of apply_filters() based on its optional preceding docblock.
+ */
+
+declare(strict_types=1);
+
+namespace SzepeViktor\PHPStan\WordPress;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\PhpDoc\PhpDocStringResolver;
+use PHPStan\PhpDoc\TypeNodeResolver;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\FileTypeMapper;
+use PHPStan\Type\Type;
+use PHPStan\Type\MixedType;
+
+class ApplyFiltersDynamicFunctionReturnTypeExtension implements \PHPStan\Type\DynamicFunctionReturnTypeExtension
+{
+    protected FileTypeMapper $fileTypeMapper;
+    protected PhpDocStringResolver $phpDocStringResolver;
+    protected TypeNodeResolver $typeNodeResolver;
+
+    public function __construct(
+        FileTypeMapper $fileTypeMapper,
+        PhpDocStringResolver $phpDocStringResolver,
+        TypeNodeResolver $typeNodeResolver
+    ) {
+        $this->fileTypeMapper = $fileTypeMapper;
+        $this->phpDocStringResolver = $phpDocStringResolver;
+        $this->typeNodeResolver = $typeNodeResolver;
+    }
+
+    public function isFunctionSupported(FunctionReflection $functionReflection): bool
+    {
+        return in_array(
+            $functionReflection->getName(),
+            [
+                'apply_filters',
+                'apply_filters_deprecated',
+                'apply_filters_ref_array',
+            ],
+            true
+        );
+    }
+
+    // phpcs:ignore SlevomatCodingStandard.Functions.UnusedParameter
+    public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+    {
+        $default = new MixedType();
+        $comment = self::getNullableNodeComment($functionCall);
+
+        if ($comment === null) {
+            return $default;
+        }
+
+        // Fetch the docblock contents and resolve it to a PhpDocNode.
+        $code = $comment->getText();
+        $doc = $this->phpDocStringResolver->resolve($code);
```

Comment:
```suggestion
        /** @phpstan-ignore-next-line */
        $doc = $this->phpDocStringResolver->resolve($code);
```

> Calling PHPStan\PhpDoc\PhpDocStringResolver::resolve() is not covered by backward compatibility promise.

Reply:
Yeah I need to look into whether there's an API method for this. I don't want to ignore it because it might break.

Reply:
PHPDoc parsing seem to be me as internal stuff...